pinot-core/src/main/java/org/apache/pinot/server/realtime/ControllerLeaderLocator.java

```diff
@@ -80,75 +87,127 @@ public static ControllerLeaderLocator getInstance() {
    */
   public synchronized Pair<String, Integer> getControllerLeader(String rawTableName) {
-    if (!_cachedControllerLeaderInvalid) {
-      return _controllerLeaderHostPort;
+    int partitionId = LeadControllerUtils.getPartitionIdForTable(rawTableName);
+    if (_cachedControllerLeaderValid) {
+      return _cachedControllerLeaderMap.get(partitionId);
     }
 
-    Pair<String, Integer> leaderForTable = getLeaderForTable(rawTableName);
-    if (leaderForTable == null) {
-      LOGGER.warn("Failed to find a leader for Table: {}", rawTableName);
-      _cachedControllerLeaderInvalid = true;
-      return null;
-    } else {
-      _controllerLeaderHostPort = leaderForTable;
-      _cachedControllerLeaderInvalid = false;
-      LOGGER.info("Setting controller leader to be {}:{}", _controllerLeaderHostPort.getFirst(),
-          _controllerLeaderHostPort.getSecond());
-      return _controllerLeaderHostPort;
-    }
+    // No controller leader cached, fetches a fresh copy of external view and then gets the leader for the given table.
+    refreshControllerLeaderMap();
+    return _cachedControllerLeaderValid ? _cachedControllerLeaderMap.get(partitionId) : null;
   }
 
   /**
-   * Firstly checks whether lead controller resource has been enabled or not.
-   * If yes, use this as the leader for realtime segment completion once partition leader exists.
-   * Otherwise, try to use Helix leader.
-   * @param rawTableName table name without type.
-   * @return the controller leader id with hostname and port for this table, e.g. localhost_9000
+   * Checks whether lead controller resource has been enabled or not.
+   * If yes, updates lead controller pairs from the external view of lead controller resource.
+   * Otherwise, updates lead controller pairs from Helix cluster leader.
    */
-  private Pair<String, Integer> getLeaderForTable(String rawTableName) {
+  private void refreshControllerLeaderMap() {
     // Checks whether lead controller resource has been enabled or not.
     if (isLeadControllerResourceEnabled()) {
-      // Gets leader from lead controller resource.
-      return getLeaderFromLeadControllerResource(rawTableName);
+      refreshControllerLeaderMapFromLeadControllerResource();
     } else {
-      // Gets Helix leader to be the leader to this table, otherwise returns null.
-      return getHelixClusterLeader();
+      refreshControllerLeaderMapFromHelixClusterLeader();
     }
   }
 
   /**
-   * Checks whether lead controller resource is enabled or not. The switch is in resource config.
+   * Updates lead controller pairs from the external view of lead controller resource.
    */
-  private boolean isLeadControllerResourceEnabled() {
-    BaseDataAccessor<ZNRecord> dataAccessor = _helixManager.getHelixDataAccessor().getBaseDataAccessor();
-    Stat stat = new Stat();
+  private void refreshControllerLeaderMapFromLeadControllerResource() {
+    boolean refreshSucceeded = false;
```

Comment:
We don't need this extra refreshSucceeded. Just set _cachedControllerLeaderValid to true as the last statement

Reply:
Updated.